subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/PathUtil.java

```diff
@@ -119,28 +196,94 @@ public static int sizeOfCommonPrefix(String path1, String path2, int offset) {
         return lastSeparator;
     }
 
-    public static int compareWithCommonPrefix(String path1, String path2, int offset) {
+    /**
+     * Compares based on the first prefix of two paths.
+     *
+     * The paths must not start with a separator, taking into account the offset
+     *
+     * @return 0 when the two paths have a common prefix, and the comparison of the first prefix if not.
+     */
+    public static int compareWithCommonPrefix(String path1, String path2, int offset, CaseSensitivity caseSensitivity) {
         int maxPos = Math.min(path1.length(), path2.length() - offset);
+        int accumulatedValue = 0;
         for (int pos = 0; pos < maxPos; pos++) {
             char charInPath1 = path1.charAt(pos);
             char charInPath2 = path2.charAt(pos + offset);
-            int comparedChars = compareChars(charInPath1, charInPath2);
+            int comparedChars = compareCharsIgnoringCase(charInPath1, charInPath2);
             if (comparedChars != 0) {
                 return comparedChars;
             }
+            accumulatedValue = computeCombinedCompare(accumulatedValue, charInPath1, charInPath2, caseSensitivity == CASE_SENSITIVE);
             if (isFileSeparator(charInPath1)) {
                 if (pos > 0) {
-                    return 0;
+                    return accumulatedValue;
                 }
             }
         }
         if (path1.length() == path2.length() - offset) {
-            return 0;
+            return accumulatedValue;
         }
         if (path1.length() > path2.length() - offset) {
-            return isFileSeparator(path1.charAt(maxPos)) ? 0 : 1;
+            return isFileSeparator(path1.charAt(maxPos)) ? accumulatedValue : 1;
+        }
+        return isFileSeparator(path2.charAt(maxPos + offset)) ? accumulatedValue : -1;
+    }
+
+    /**
+     * Determines whether a path starting at an offset has the given prefix,
```

Comment:
It would be good to explain how we use this `String path` + `int offset` thing for optimization. (It would be even better to capture this behavior in a class of its own, and perhaps move many of the methods defined here in `PathUtil` to that class.)

Reply:
I introduced a separate class in https://github.com/gradle/gradle/pull/11378. I would prefer to merge this PR, and then do the extraction. WDYT?